mitiq/ddd/insertion.py

```diff
@@ -0,0 +1,58 @@
+# Copyright (C) 2022 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Tools to determine slack windows in circuits and to insert DDD sequences."""
+
+import numpy as np
+
+
+def _validate_integer_matrix(mask: np.ndarray) -> None:
+    """Ensures the input is a NumPy 2d array with integer elements."""
+    if not isinstance(mask, np.ndarray):
+        raise TypeError("The input matrix must be a numpy.ndarray object.")
+    if not np.issubdtype(mask.dtype.type, int):
+        raise TypeError("The input matrix must have integer elements.")
+    if len(mask.shape) != 2:
+        raise ValueError("The input must be a 2-dimensional array.")
+
+
+def get_slack_matrix_from_circuit_mask(mask: np.ndarray) -> np.ndarray:
+    """Given a circuit mask matrix A, e.g. the output of get_circuit_mask(),
+    returns a slack matrix B, where B_{i,j} = t if the position A{i,j} is the
+    initial element of a sequence of t zeros (from left to right).
+
+    Args:
+        mask: The mask matrix of a quantum circuit.
+
+    Returns: The matrix of slack lengths.
+    """
+    _validate_integer_matrix(mask)
+    if not (mask**2 == mask).all():
+        raise ValueError("The input matrix elements must be 0 or 1.")
+
+    num_raws, num_cols = mask.shape
+    slack_matrix = np.zeros((num_raws, num_cols))
```

Comment:
Do we want to specify the slack matrix dtype as int? i.e. `slack_matrix = np.zeros((num_raws, num_cols), int)`

Reply:
Good point! Done here: [8586900](https://github.com/unitaryfund/mitiq/pull/1177/commits/858690092e75deb721fd3bd58ea15396d5a61e36)